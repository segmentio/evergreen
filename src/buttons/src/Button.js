import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import { Text } from '../../typography'
import { IconMap, IconAim } from '../../icons'
import {
  getBorderRadiusForControlHeight,
  getTextStyleForControlHeight,
  getIconSizeForControlHeight
} from '../../shared-styles'
import { Spinner } from '../../spinner'
import ButtonAppearances from './styles/ButtonAppearances'

export default class Button extends PureComponent {
  static propTypes = {
    /**
     * Composes the Text component as the base.
     */
    ...Text.propTypes,

    /**
     * The appearance of the button.
     */
    appearance: PropTypes.oneOf(Object.keys(ButtonAppearances)).isRequired,

    /**
     * When true, show a loading spinner before the children.
     * This also disables the button.
     */
    isLoading: PropTypes.bool,

    /**
     * Forcefully set the active state of a button.
     * Useful in conjuction with a Popover.
     */
    isActive: PropTypes.bool,

    /**
     * Sets an icon before the text. Can be any icon from Evergreen.
     */
    iconBefore: PropTypes.oneOf(Object.keys(IconMap)),

    /**
     * The aim of the left icon. Not a big use case for this.
     */
    iconBeforeAim: PropTypes.oneOf(Object.keys(IconAim)),

    /**
     * Sets an icon after the text. Can be any icon from Evergreen.
     */
    iconAfter: PropTypes.oneOf(Object.keys(IconMap)),

    /**
     * The aim of the right icon. Useful to aim a triangle down.
     */
    iconAfterAim: PropTypes.oneOf(Object.keys(IconAim)),

    /**
     * When true, the button is disabled.
     * isLoading also sets the button to disabled.
     */
    disabled: PropTypes.bool
  }

  static defaultProps = {
    appearance: 'default',
    isActive: false,
    iconBeforeAim: 'none',
    iconAfterAim: 'none',
    height: 32,
    paddingTop: 0,
    paddingBottom: 0
  }

  static styles = {
    position: 'relative',
    fontFamily: 'ui',
    fontWeight: 500,
    display: 'inline-flex',
    alignItems: 'center',
    flexWrap: 'nowrap'
  }

  render() {
    const {
      css,
      height,
      isActive,
      children,
      disabled,
      appearance,
      isLoading,

      // Paddings
      paddingRight,
      paddingLeft,
      paddingTop,
      paddingBottom,

      // Icons
      iconBefore: iconBeforeKey,
      iconBeforeAim,
      iconAfter: iconAfterKey,
      iconAfterAim,

      ...props
    } = this.props
    const appearanceStyle = ButtonAppearances[appearance]
    const textStyle = getTextStyleForControlHeight({ height })
    const borderRadius = getBorderRadiusForControlHeight({ height })
    const iconHeight = height - 4
    const iconSize = getIconSizeForControlHeight({ height: iconHeight })

    const pr =
      paddingRight !== undefined ? paddingRight : Math.round(height / 2) // eslint-disable-line no-negated-condition
    const pl = paddingLeft !== undefined ? paddingLeft : Math.round(height / 2) // eslint-disable-line no-negated-condition

    let iconBefore
    if (iconBeforeKey) {
      iconBefore = React.createElement(IconMap[iconBeforeKey], {
        aim: iconBeforeAim,
        iconSize,
        color: 'inherit',
        size: iconHeight,
        marginLeft: -Math.round(pl * 0.6)
      })
    }

    let iconAfter
    if (iconAfterKey) {
      iconAfter = React.createElement(IconMap[iconAfterKey], {
        aim: iconAfterAim,
        iconSize,
        color: 'inherit',
        size: iconHeight,
        marginRight: -Math.round(pl * 0.6)
      })
    }

    return (
      <Text
        is="button"
        borderTopRightRadius={borderRadius}
        borderBottomRightRadius={borderRadius}
        borderTopLeftRadius={borderRadius}
        borderBottomLeftRadius={borderRadius}
        paddingTop={paddingTop}
        paddingBottom={paddingBottom}
        paddingRight={pr}
        paddingLeft={pl}
        margin={0} // Removes weird margins in Safari
        {...textStyle}
        color={null} // Prevent the Text color overriding the glamor appearanceStyle color
        css={{
          ...appearanceStyle,
          ...css
        }}
        height={height}
        lineHeight={`${height}px`}
        {...(isActive ? { 'data-active': true } : {})}
        {...Button.styles}
        {...props}
        disabled={disabled || isLoading}
      >
        {isLoading && (
          <Spinner
            marginLeft={-Math.round(height / 8)}
            marginRight={Math.round(height / 4)}
            size={Math.round(height / 2)}
          />
        )}
        {iconBefore || null}
        <span>{children}</span>
        {iconAfter || null}
      </Text>
    )
  }
}
