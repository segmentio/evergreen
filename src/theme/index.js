export { default as defaultTheme } from './src/default-theme'
export {
  default as ThemeContext,
  ThemeProvider,
  ThemeConsumer
} from './src/ThemeContext'
export { default as withTheme } from './src/withTheme'
export { default as useTheme } from './src/useTheme'
