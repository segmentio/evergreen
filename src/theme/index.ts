export { default as ThemeContext, getThemeContext, ThemeProvider, ThemeConsumer } from './src/ThemeContext'
export { default as withTheme } from './src/withTheme'
export { default as useTheme } from './src/useTheme'
export { get, getValue, mergeTheme, resolveThemeTokens } from './src/theme-tools'
