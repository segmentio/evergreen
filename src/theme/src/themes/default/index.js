/**
 * Theme
 * ---
 * The theme object is used to style Evergreen.
 * It is passed into the  `<ThemeProvider theme={theme} />`.
 * ----
 * You can use this as a template for your own themes.
 */

/**
 * Foundational Styles.
 * ---
 * The following properties are NOT REQUIRED by Evergreen.
 * It's exposed for convenience and documentation.
 */
import {
  colors,
  elevations,
  fills,
  palette,
  scales,
  tokens
} from './foundational-styles'

/**
 * Typography.
 * ---
 * The following properties are NOT REQUIRED by Evergreen.
 * It's exposed for convenience and documentation.
 */
import { headings, text, fontFamilies, paragraph } from './typography'

/**
 * Component Specific.
 * ---
 * These ARE REQUIRED for Evergreen to work.
 */
import {
  getTextDropdownButtonClassName,
  getTableCellClassName,
  getRowClassName,
  getMenuItemClassName,
  buttons,
  inputs
} from './component-specific'

/**
 * Theme Helpers.
 * ---
 * These ARE REQUIRED for Evergreen to work.
 */
import { getElevation } from './theme-helpers'

export default {
  // Foundational Styles.
  colors,
  elevations,
  fills,
  palette,
  scales,
  tokens,

  getTextDropdownButtonClassName,
  getTableCellClassName,
  getRowClassName,
  getMenuItemClassName,

  getElevation,

  inputs,
  buttons,

  typography: {
    headings,
    text,
    fontFamilies,
    paragraph
  }
}
