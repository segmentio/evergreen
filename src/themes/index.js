export { default as defaultTheme } from './default'
