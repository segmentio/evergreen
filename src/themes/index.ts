export { default as defaultTheme } from './default'
export type {
  Components,
  DefaultTheme,
  DefaultThemeAppearances,
  DefaultThemeColors,
  DefaultThemeColor,
  DefaultThemeFills,
  DefaultThemeIntents,
  DefaultThemeIntent,
  DefaultThemePseudoSelectors,
  DefaultThemeSizes,
} from './default'
export { default as classicTheme } from './classic'
export { default as deprecatedDefaultTheme } from './deprecated'
