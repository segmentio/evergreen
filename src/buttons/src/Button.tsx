import React, { PureComponent, ReactNode } from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'
import { Text } from '../../typography'
import { Icon, IconName } from '../../icon'
import { Spinner } from '../../spinner'
import { withTheme, Theme } from '../../theme'

type Intent = 'none' | 'success' | 'warning' | 'danger'
type Appearance = 'default' | 'minimal' | 'primary'

interface ButtonProps extends React.ComponentProps<typeof Text> {
  /**
   * The intent of the button.
   */
  intent: Intent

  /**
   * The appearance of the button.
   */
  appearance: Appearance

  /**
   * Sets the height and font-size of the button.
   */
  height: number

  /**
   * When true, show a loading spinner before the children.
   * This also disables the button.
   */
  isLoading?: boolean

  /**
   * Forcefully set the active state of a button.
   * Useful in conjuction with a Popover.
   */
  isActive?: boolean

  /**
   * Sets an icon before the text. Can be any icon from Evergreen.
   */
  iconBefore?: IconName

  /**
   * Sets an icon after the text. Can be any icon from Evergreen.
   */
  iconAfter?: IconName

  /**
   * When true, the button is disabled.
   * isLoading also sets the button to disabled.
   */
  disabled?: boolean

  /**
   * Theme provided by ThemeProvider.
   */
  theme: Theme
}

class Button extends PureComponent<ButtonProps> {
  static propTypes = {
    intent: PropTypes.oneOf(['none', 'success', 'warning', 'danger'])
      .isRequired as PropTypes.Validator<Intent>,
    appearance: PropTypes.oneOf(['default', 'minimal', 'primary'])
      .isRequired as PropTypes.Validator<Appearance>,
    height: PropTypes.number.isRequired,
    isLoading: PropTypes.bool,
    isActive: PropTypes.bool,
    iconBefore: PropTypes.string as PropTypes.Requireable<IconName>,
    iconAfter: PropTypes.string as PropTypes.Requireable<IconName>,
    disabled: PropTypes.bool,
    theme: PropTypes.object.isRequired as PropTypes.Validator<Theme>
  }

  static defaultProps = {
    appearance: 'default' as const,
    height: 32,
    intent: 'none' as const,
    isActive: false,
    paddingBottom: 0,
    paddingTop: 0
  }

  static styles = {
    position: 'relative',
    fontFamily: 'ui',
    fontWeight: 500,
    display: 'inline-flex',
    alignItems: 'center',
    flexWrap: 'nowrap'
  }

  render() {
    const {
      theme,
      className,

      intent,
      height,
      isActive,
      children,
      disabled,
      appearance,
      isLoading,

      // Paddings
      paddingRight,
      paddingLeft,
      paddingTop,
      paddingBottom,

      // Icons
      iconBefore: iconBeforeKey,
      iconAfter: iconAfterKey,

      ...props
    } = this.props

    const themedClassName = theme.getButtonClassName(appearance, intent)
    const textSize = theme.getTextSizeForControlHeight(height)

    const borderRadius = theme.getBorderRadiusForControlHeight(height)
    const iconSize = theme.getIconSizeForButton(height)

    const pr =
      typeof paddingRight === 'number' ? paddingRight : Math.round(height / 2)
    const pl =
      typeof paddingLeft === 'number' ? paddingLeft : Math.round(height / 2)

    let iconBefore: ReactNode | undefined
    if (iconBeforeKey) {
      iconBefore = (
        <Icon
          icon={iconBeforeKey}
          size={iconSize}
          marginLeft={-Math.round(pl * 0.2)}
          marginRight={Math.round(iconSize * 0.7)}
        />
      )
    }

    let iconAfter: ReactNode | undefined
    if (iconAfterKey) {
      iconAfter = (
        <Icon
          icon={iconAfterKey}
          size={iconSize}
          marginRight={-Math.round(pl * 0.2)}
          marginLeft={Math.round(iconSize * 0.7)}
        />
      )
    }

    return (
      <Text
        is="button"
        className={cx(themedClassName, className)}
        borderTopRightRadius={borderRadius}
        borderBottomRightRadius={borderRadius}
        borderTopLeftRadius={borderRadius}
        borderBottomLeftRadius={borderRadius}
        paddingTop={paddingTop}
        paddingBottom={paddingBottom}
        paddingRight={pr}
        paddingLeft={pl}
        marginLeft={0} // Removes weird margins in Safari
        marginRight={0} // Removes weird margins in Safari
        marginTop={0} // Removes weird margins in Safari
        marginBottom={0} // Removes weird margins in Safari
        size={textSize}
        color={null} // Prevent the Text color overriding the glamor appearanceStyle color
        height={height}
        lineHeight={`${height}px`}
        {...(isActive ? { 'data-active': true } : {})}
        {...Button.styles}
        {...props}
        disabled={disabled || isLoading}
      >
        {isLoading && (
          <Spinner
            marginLeft={-Math.round(height / 8)}
            marginRight={Math.round(height / 4)}
            size={Math.round(height / 2)}
          />
        )}
        {iconBefore || null}
        {children}
        {iconAfter || null}
      </Text>
    )
  }
}

export default withTheme(Button)
