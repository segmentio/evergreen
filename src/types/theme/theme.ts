import { Components } from '../../themes'
import { Color } from './color'
import { ComponentStyles } from './component-styles'
import { Fill } from './fill'
import { FontFamilies } from './font-families'
import { FontWeights } from './font-weights'
import { Intent } from './intent'
import { LettingSpacings } from './letter-spacings'
import { ZIndices } from './z-indices'

export interface Theme<TComponents extends Components = Components> {
  colors: { [color: string]: Color<string | { [group: string]: Color }> }
  fills: { [fill: string]: Fill }
  intents: { [intent: string]: Intent }
  fontFamilies: FontFamilies
  radii: string[]
  shadows: string[] & { focusRing: string }
  fontSizes: string[]
  fontWeights: FontWeights
  letterSpacings: LettingSpacings
  lineHeights: string[]
  zIndices: ZIndices
  components: Partial<ComponentStyles<TComponents>>
}
