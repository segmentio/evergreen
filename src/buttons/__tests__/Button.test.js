import React from 'react'
import render from 'react-test-renderer'
import { ThemeProvider } from '../../theme'
import { classicTheme, defaultTheme } from '../../themes'
import Button from '../src/Button'

describe.each([
  ['default', defaultTheme],
  ['classic', classicTheme]
])('<Button /> % %s', (_, theme) => {
  it('snapshots with the rendered output', () => {
    const component = (
      <ThemeProvider value={theme}>
        <Button />
      </ThemeProvider>
    )
    const tree = render.create(component).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
