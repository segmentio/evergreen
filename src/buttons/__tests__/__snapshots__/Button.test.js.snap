// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> % classic snapshots with the rendered output 1`] = `
<button
  className="css-15cxcpy ub-pst_relative ub-f-wght_500 ub-dspl_inline-flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_0 ub-b-lft_0 ub-b-rgt_0 ub-b-top_0 ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bblr_3px ub-bbrr_3px ub-btlr_3px ub-btrr_3px ub-color_425A70 ub-h_32px ub-min-w_32px ub-fnt-sze_12px ub-ln-ht_32px ub-pl_16px ub-pr_16px ub-bg-clr_white ub-bg-img_zk1eut ub-bs_ruftdn ub-box-szg_border-box"
/>
`;

exports[`<Button /> % default snapshots with the rendered output 1`] = `
<button
  className="css-12b25td ub-pst_relative ub-f-wght_500 ub-dspl_inline-flex ub-algn-itms_center ub-flx-wrap_nowrap ub-just-cnt_center ub-txt-deco_none ub-ver-algn_middle ub-b-btm_1px-solid-c1c4d6 ub-b-lft_1px-solid-c1c4d6 ub-b-rgt_1px-solid-c1c4d6 ub-b-top_1px-solid-c1c4d6 ub-otln_iu2jf4 ub-usr-slct_none ub-crsr_pointer ub-wht-spc_nowrap ub-fnt-fam_b77syt ub-bblr_4px ub-bbrr_4px ub-btlr_4px ub-btrr_4px ub-color_474d66 ub-tstn_n1akt6 ub-h_32px ub-min-w_32px ub-fnt-sze_12px ub-ln-ht_32px ub-pl_16px ub-pr_16px ub-bg-clr_white ub-box-szg_border-box"
/>
`;
